src/mqtt/buffer_ref.h

```diff
@@ -0,0 +1,278 @@
+	 */
+	buffer_ref& operator=(const char* cstr) {
+		static_assert(sizeof(char) == sizeof(T), "can only use C arr with char or byte buffers");
+		data_.reset(new blob(reinterpret_cast<const value_type*>(cstr), strlen(cstr)));
+		return *this;
+	}
+	/**
+	 * Copy another type of buffer reference to this one.
+	 * This can copy a buffer of different types, provided that the size of
+	 * the data elements are the same. This is typically used to convert
+	 * from char to byte, where the data is the same, but the interpretation
+	 * is different. Note that this copies the underlying buffer.
+	 * @param rhs A reference to a different type of buffer.
+	 * @return A reference to this object.
+	 */
+	template <typename OT>
+	buffer_ref& operator=(const buffer_ref<OT>& rhs) {
+		static_assert(sizeof(OT) == sizeof(T), "Can only assign buffers if values the same size");
+		data_.reset(new blob(reinterpret_cast<const value_type*>(rhs.data()), rhs.size()));
+		return *this;
+	}
+	/**
+	 * Clears the reference to nil.
+	 */
+	void reset() { data_.reset(); }
+	/**
+	 * Get a const pointer to the data buffer.
+	 * @return A pointer to the data buffer.
+	 */
+	const value_type* data() const { return data_ ? data_->data() : nullptr; }
+	/**
+	 * Get the size of the data buffer.
+	 * @return The size of the data buffer.
+	 */
+	size_t size() const { return data_ ? data_->size() : 0; }
+	/**
+	 * Get the size of the data buffer.
+	 * @return The size of the data buffer.
+	 */
+	size_t length() const { return data_ ? data_->length() : 0; }
+	/**
+	 * Get the data buffer as a string.
+	 * @return The data buffer as a string.
+	 */
+	//blob to_string() const { return data_ ? (*data_) : blob(); }
+	const blob& to_string() const { return *data_; }
+	/**
+	 * Get the data buffer as NUL-terminated C string.
+	 * Note that the reference must be set to call this function.
+	 * @return The data buffer as a string.
+	 */
+	const char* c_str() const { return data_->c_str(); }
+	/**
+	 * Gets a shared pointer to the (const) data buffer.
+	 * @return A shared pointer to the (const) data buffer.
+	 */
+	const pointer_type& ptr() const { return data_; }
+	/**
+	 * Gets elemental access to the data buffer (read only)
+	 * @param i The index into the buffer.
+	 * @return The value at the specified index.
+	 */
+	const value_type& operator[](size_t i) const { return (*data_)[i]; }
+};
+
+template <typename T>
+std::ostream& operator<<(std::ostream& os, const buffer_ref<T>& buf) {
+	if (!buf.empty())
+		os.write(buf.data(), buf.size());
+	return os;
+}
+
+/////////////////////////////////////////////////////////////////////////////
+
+/**
+ * A refernce to a text buffer.
+ */
+using string_ref = buffer_ref<char>;
+
+/**
+ * A reference to a binary buffer.
+ * Note that we're using char for the underlying data type to allow
+ * efficient moves to and from std::string's. Using a separate type
+ * indicates that the data may be arbitrary binary.
+ */
+using binary_ref = buffer_ref<char>;
```

Comment:
Oh, sorry, I wasn't being clear. This would be the underlying type that gets wrapped in the binary_ref. So a string_ref would be a shared pointer to a string, and a binary_ref would be a shred pointer to one of these things.

But I'm probably over thinking it at this point. Using a std::string for either is probably fine.